feat(use-case-b): Access KFP from user pods

orfeas-k opened the pull request:
Migrate use case b to enable user pods to access KFP pods. This is showcased by enabling pipelines to run successfully from the KFP UI.
Close canonical/bundle-kubeflow#1286

kimwnasptd commented on use-case-b-kfp-from-user-pods-access/README.md:
````diff
@@ -0,0 +1,142 @@
+2. Add the user namespace to ambient mesh and disable sidecar injection
+```sh
+kubectl label ns kubeflow-user-example-com istio.io/dataplane-mode=ambient
+kubectl label ns kubeflow-user-example-com istio-injection-
````
**suggestion(blocking)**: Let's fix the typo here, as the label for now is `istio-injection-`

orfeas-k replied:
This is not a typo, running this command removes the `istio-injection` label.
```sh
$ kgns demo -o yaml           
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: "2025-07-18T10:22:31Z"
  labels:
    kubernetes.io/metadata.name: demo
    owner: orfeas
  name: demo
  resourceVersion: "297515"
  uid: 7ac50e8d-5b72-45cd-9896-89485538e92a
spec:
  finalizers:
  - kubernetes
status:
  phase: Active

$ k label ns demo owner-      
namespace/demo unlabeled

$ kgns demo -o yaml     
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: "2025-07-18T10:22:31Z"
  labels:
    kubernetes.io/metadata.name: demo
  name: demo
  resourceVersion: "297622"
  uid: 7ac50e8d-5b72-45cd-9896-89485538e92a
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

kimwnasptd replied:
(for some reason I can't resolve the comment. ACK tho)

kimwnasptd commented on use-case-b-kfp-from-user-pods-access/kfp-ap-l7.yaml:
```diff
@@ -0,0 +1,52 @@
+    - key: request.headers[kubeflow-userid]
+      notValues:
+      - '*'
+  targetRefs:
+  - kind: Service
+    name: ml-pipeline
+    group: ""
```
**questions**:

What if a (malicious) Pod, with annotations/labels that exclude it from the mesh, sets `kubeflow-userid` header with any value and sends a request to the:
1. `ml-pipeline` service?
2. `ml-pipeline` Pod IP?

Should we have a way to force the KFP Pods/Services to only accept traffic from the waypoint?

orfeas-k replied:
In [use case a](https://github.com/orfeas-k/kubeflow-istio-ambient-demo/tree/main/use-case-a-web-application-access), two deny-all-allow-nothing policies were introduced. I just noticed actually that this wasn't thoroughly documented, although it is mentioned in the relevant [issue](https://github.com/canonical/bundle-kubeflow/issues/1285#issuecomment-3056440427:~:text=All%20enforcement%20places%20are%20protected%20via%20with%20a%20default%20deny%20(allow%20nothing)%20policy). This means that traffic both on L4 and L7 should not pass, unless there is a rule allowing it. Actually, even without that allow nothing, since we have the `kfp-ap-l4-waypoint.yaml` policies attached to those services, traffic that doesn't match will be refused.

orfeas-k replied:
The destination pod ztunnel enforces L4 authPolicies https://istio.io/latest/docs/ambient/usage/add-workloads/#pods-outside-the-mesh:~:text=while%20the%20destination%20pod%E2%80%99s%20ztunnel%20enforces%20any%20L4%20policy%20to%20control%20whether%20traffic%20should%20be%20allowed%20or%20denied.

kimwnasptd replied:
Ah OK, this makes sense.

(consider this one resolved)

kimwnasptd reviewed (commented):
Really nice work @orfeas-k!

kimwnasptd commented:
LGTM!

orfeas-k committed:
feat(use-case-b): Access KFP from user pods

Migrate use case b to enable user pods to access KFP pods. This is
showcased by enabling pipelines to run successfully from the KFP UI.
